src/sage/rings/species.py

```diff
@@ -0,0 +1,1752 @@
+            sage: G = PermutationGroup([[(1,2)], [(3,4)]]); G
+            Permutation Group with generators [(3,4), (1,2)]
+            sage: G.disjoint_direct_product_decomposition()
+            {{1, 2}, {3, 4}}
+            sage: G in A
+            False
+        """
+        if parent(x) == self:
+            return True
+        G, pi = None, None
+        if isinstance(x, PermutationGroup_generic):
+            if self._arity == 1:
+                G = x
+                pi = {1: G.domain()}
+            else:
+                raise ValueError("the assignment of sorts to the domain elements must be provided")
+        else:
+            G, pi = x
+        if not isinstance(G, PermutationGroup_generic):
+            raise ValueError(f"{G} must be a permutation group")
+        if not set(pi.keys()).issubset(range(1, self._arity + 1)):
+            raise ValueError(f"keys of pi (={pi.keys()}) must be in the range [1, {self._arity}]")
+        if sum(len(p) for p in pi.values()) != len(G.domain()) or set(chain.from_iterable(pi.values())) != set(G.domain()):
+            raise ValueError(f"values of pi (={pi.values()}) must partition the domain of G (={G.domain()})")
+        for orbit in G.orbits():
+            if not any(set(orbit).issubset(p) for p in pi.values()):
+                raise ValueError(f"All elements of orbit {orbit} must have the same sort")
+        return len(G.disjoint_direct_product_decomposition()) <= 1
+
+    def _repr_(self):
+        r"""
+        Return a string representation of ``self``.
+
+        TESTS::
+
+            sage: AtomicSpecies("X")
+            Atomic species in X
+            sage: AtomicSpecies("X, Y")
+            Atomic species in X, Y
+        """
+        if len(self._names) == 1:
+            return f"Atomic species in {self._names[0]}"
+        return f"Atomic species in {', '.join(self._names)}"
+
+    Element = AtomicSpeciesElement
+
+
+class MolecularSpecies(IndexedFreeAbelianMonoid, ElementCache):
+    @staticmethod
+    def __classcall__(cls, indices, prefix=None, **kwds):
+        return super(IndexedMonoid, cls).__classcall__(cls, indices, prefix, **kwds)
+
+    def __init__(self, indices, prefix=None, **kwds):
+        r"""
+        Infinite set of multivariate molecular species.
+
+        INPUT:
+
+        - ``indices`` -- the underlying set of atomic species indexing the monoid
+
+        TESTS::
+
+            sage: P1 = PolynomialSpecies(ZZ, "X")
+            sage: P2 = PolynomialSpecies(ZZ, ["X", "Y"])
+            sage: TestSuite(P1._indices).run(skip="_test_graded_components")
+            sage: TestSuite(P2._indices).run(skip="_test_graded_components")
+        """
+        category = Monoids() & SetsWithGrading().Infinite()
+        IndexedFreeAbelianMonoid.__init__(self, indices, prefix=prefix, category=category, **kwds)
+        ElementCache.__init__(self)
+        self._arity = indices._arity
+
+    def _project(self, G, pi, part):
```

Comment:
This does not depend on `self` currently.  Maybe it should really be `def _project(self, part)`?